src/ConstructiveSolidGeometry/SurfacePrimitives/EllipticalSurface.jl

```diff
@@ -18,12 +18,44 @@ Surface primitive describing circular bases, e.g. the top or bottom base of a [`
 * `origin::CartesianPoint{T}`: The position of the center of the `EllipticalSurface`.
 * `rotation::SMatrix{3,3,T,9}`: Matrix that describes a rotation of the `EllipticalSurface` around its `origin`.
 """
-@with_kw struct EllipticalSurface{T,TR,TP<:Union{Nothing,T}} <: AbstractPlanarSurfacePrimitive{T}
-    r::TR = 1
-    φ::TP = nothing
+struct EllipticalSurface{T,TR,TP<:Union{Nothing,T}} <: AbstractPlanarSurfacePrimitive{T}
+    r::TR
+    φ::TP 
 
-    origin::CartesianPoint{T} = zero(CartesianPoint{T})
-    rotation::SMatrix{3,3,T,9} = one(SMatrix{3, 3, T, 9})
+    origin::CartesianPoint{T}
+    rotation::SMatrix{3,3,T,9} 
+end
+
+#Type conversion happens here
+function EllipticalSurface{T}(r, φ, origin, rotation) where {T}
+    _r = _csg_convert_args(T, r)
+    (_φ, _rotation) = _handle_phi(_csg_convert_args(T, φ), rotation)
+    EllipticalSurface{T,typeof(_r),typeof(_φ)}(_r, _φ, origin, _rotation)
+end
+
+#Type promotion happens here
+function EllipticalSurface(::Nothing,r::TR, φ::TP, origin::PT, rotation::ROT) where {TR, TP, PT, ROT}
+    eltypes = _csg_get_promoted_eltype.((TR, TP, PT, ROT))
+    T = float(promote_type(eltypes...))
+    EllipticalSurface{T}(r, φ, origin, rotation)
+end
+
+function EllipticalSurface(;
+    # define default parameters as Int to not influence type promotion
+    r = 1, 
+    φ = nothing,
+    origin = zero(CartesianPoint{Int}), 
+    rotation = one(SMatrix{3, 3, Int, 9}))
+    EllipticalSurface(nothing,r, φ, origin, rotation)#First argument is ugly workaround to distinguish promotion function and struct
```

Comment:
You also write this in a comment here that this is an ugly workaround.
I guess, the definition of the `struct` already comes with a constructor for `EllipticalSurface`, even when no parametric types are passed. You might wanna specify the arguments up to a point where Julia knows which method to use (if the variables are not yet type promoted).